# Outline of DESeq2 analysis.
# Modify for particular datasets.

library(data.table)
library(DESeq2)
library(gplots)
library(RColorBrewer)

# Root of data directory
data_dir = "DataProjects/phospholamban-cardiomyocytes/RNA-seq"

# Load data
# --------------------------------------------------

# Load gene expression count matrices generated by tuxedo2 
# .cvs files collated by prepDE.py
count_mat = fread(file.path(data_dir, "ribo-zero/results/gene_count_matrix.csv"))

# Parse gene count matrix
row_ids = count_mat[[1]]
count_mat = as.matrix(count_mat[, -1])
rownames(count_mat) = row_ids

# Load gene annotation table
gene_annot = fread(file.path(data_dir, "ribo-zero/results/gene_annot.csv"))

# Make sure that IDs in annotation table match matrix rows
if (!all(gene_annot[,1] == rownames(count_mat))) {
	stop("Transcript ID mismatch.")
}

# Load sample info
pheno = fread(file.path(data_dir, "sample_info.tsv"))

# Parse sample annotation table
pheno$id = sapply(strsplit(pheno$file_name, "[.]"), function(x) x[1])
pheno = pheno[pheno$rna_protocol == "Ribo-zero"]

# match phenotype table to columns of gene expression count matrix
pheno = pheno[match(colnames(count_mat), pheno$id)]

# Exclude samples from analysis
exclude_samples = c("R9C2-3_S2_L008_R1_001", "R9C3-2_S1_L008_R1_001")
pheno = pheno[!pheno$id %in% exclude_samples]

# Remove corresponding columns from count matrix
count_mat = count_mat[, !colnames(count_mat) %in% exclude_samples]

# Check if column IDs still match
if (!all(pheno$id == colnames(count_mat))) {
	stop("Sample ID mismatch.")
}

# Differential gene expression analysis
# -----------------------------------------------

# DESeq2 differential analysis
dds = DESeqDataSetFromMatrix(
	countData=count_mat,
	colData=pheno,
	design=~patient + genotype  # linear model of patient-specific 
	# design=~sex + age_group + condition  # patient vs control, adjusting for sex and age_group
)

# Calculate 
dds = DESeq(dds, parallel=TRUE)

res = results(dds,
	independentFiltering=FALSE,
	cooksCutoff=Inf,
	parallel=TRUE)

res$gene_symbol = gene_annot$gene_name[match(rownames(res), gene_annot$gene_id)]

# Reorder based on significance
res = res[order(res$pvalue),]

# sum(res$padj < 0.2, na.rm=TRUE)
# sum(res$padj < 0.4, na.rm=TRUE)

sig_genes = res$gene_symbol[which(res$padj < 0.2)]
# sig_genes = res$gene_symbol[which(res$pval < 0.01)]
sig_genes = sig_genes[!is.na(sig_genes)]

write.table(sig_genes,
	file.path(data_dir, "ribo-zero/results/de_genes_FDR02.txt"),
	row.names=FALSE, col.names=FALSE,
	quote=FALSE)

# head(as.data.frame(res), 100)
write.table(
	res[which(res$pval < 0.05), ],
	file.path(data_dir, "ribo-zero/results/deseq2_results.tsv"),
	quote=FALSE, sep="\t",
	col.names=NA)


# Look at normalized gene expression
# -------------------------------------------

# Normalize by size factors
norm_counts = sweep(counts(dds), 2, sizeFactors(dds), "/")

# pseudo count log transform
norm_counts = log2(norm_counts + 1)

# standardize (z-scores)
norm_counts = t(scale(t(norm_counts)))

# Normalized gene expression
sig_mat = norm_counts[rownames(norm_counts) %in% rownames(res)[res$padj < 0.2],]

# Rename rows based on gene names
idx = match(rownames(sig_mat), gene_annot$gene_id)  # annotation entries
gene_names = gene_annot$gene_name[idx]
rownames(sig_mat)[!is.na(gene_names)] = gene_names[!is.na(gene_names)]

pdf("plots/sig_heatmap.pdf")
heatmap.2(sig_mat,
	col=colorRampPalette(rev(brewer.pal(9, "RdBu")))(100),
	mar=c(10, 10),
	cexCol=0.6,
	key.xlab="z-score",
	key.title="",
	trace="none")
dev.off()

# Look at selected target genes
target = fread(file.path(data_dir, "delaine_genes/R14del_fatgenes_trim.csv"))

idx = match(target$Gene, gene_annot$gene_name)
sel_mat = norm_counts[idx,]
rownames(sel_mat) = gene_annot$gene_name[idx]

pdf("plots/sel_heatmap.pdf")
heatmap.2(sel_mat,
	col=colorRampPalette(rev(brewer.pal(9, "RdBu")))(100),
	mar=c(10, 10),
	cexCol=0.6,
	key.xlab="z-score",
	key.title="",
	trace="none")
dev.off()
