# Outline of DESeq2 analysis.
# Modify for particular datasets.


library(data.table)
library(DESeq2)
library(qvalue)

# Root of data directory
data_dir = "DataProjects/phospholamban-cardiomyocytes/RNA-seq"

# Load gene expression count matrices generated by tuxedo2 
# .cvs files collated by prepDE.py
count_mat = fread(file.path(data_dir, "ribo-zero/results/gene_count_matrix.csv"))
# count_mat = rename(count_mat, c("V1"="gene_id"))  # rename for convenience

# Gene count matrix
row_ids = count_mat[[1]]
count_mat = as.matrix(count_mat[, -1])
rownames(count_mat) = row_ids


# Load gene annotation table
gene_annot = fread(file.path(data_dir, "ribo-zero/results/gene_annot.csv"))

# Make sure that IDs in annotation table match matrix rows
if (!all(gene_annot[,1] == rownames(count_mat))) {
	stop("Transcript ID mismatch.")
}

# Load sample info
pheno = fread(file.path(data_dir, "sample_info.tsv"))
pheno$id = sapply(strsplit(pheno$file_name, "[.]"), function(x) x[1])
pheno = pheno[pheno$rna_protocol == "Ribo-zero"]

# match phenotype table to 
pheno = pheno[match(colnames(count_mat), pheno$id)]

data.frame(x1=pheno$id, x2=colnames(count_mat))

# Exclude samples from analysis
exclude_samples = c("R9C2-3_S2_L008_R1_001", "R9C3-2_S1_L008_R1_001")
pheno = pheno[!pheno$id %in% exclude_samples]

# Trim count matrix
count_mat = count_mat[, !colnames(count_mat) %in% exclude_samples]

# Check if column IDs still match
if (!all(pheno$id == colnames(count_mat))) {
	stop("Sample ID mismatch.")
}


# DESeq2 differential analysis
dds = DESeqDataSetFromMatrix(
	countData=count_mat,
	colData=pheno,
	design=~patient + genotype  # linear model of patient-specific 
	# design=~sex + age_group + condition  # patient vs control, adjusting for sex and age_group
)

# Calculate 
dds = DESeq(dds, parallel=TRUE)

res = results(dds,
	independentFiltering=FALSE,
	cooksCutoff=Inf,
	parallel=TRUE)  # not sure if the parallel version works


# res$qvalue = qvalue(res$pvalue)$lfdr

res$gene_symbol = gene_annot$gene_name[match(rownames(res), gene_annot$gene_id)]

# Reorder based on significance
res = res[order(res$pvalue),]

sum(res$padj < 0.2, na.rm=TRUE)
# sum(res$padj < 0.4, na.rm=TRUE)

sig_genes = res$gene_symbol[which(res$padj < 0.2)]
# sig_genes = res$gene_symbol[which(res$padj < 0.4)]
sig_genes = sig_genes[!is.na(sig_genes)]

write.table(sig_genes, file.path(data_dir, "ribo-zero/results/de_genes.txt"),
	row.names=FALSE, col.names=FALSE,
	quote=FALSE)

head(as.data.frame(res), 100)

write.table(res, "deseq2_results.tsv", quote=FALSE, sep="\t", col.names=NA)
