library(data.table)  # for fread
library(rtracklayer)  # for readGFF
library(plyr)   # for rename

# Root of data directory
data_dir = "DataProjects/phospholamban-cardiomyocytes/RNA-seq/ribo-zero"

# Load transcript GTF tables for merged assembled transcripts generated by StringTie
read_annot = readGFF(file.path(data_dir, "merged/stringtie_merged.gtf"))

# Load gene expression count matrices generated by tuxedo2 
# .cvs files collated by prepDE.py
count_mat = fread(file.path(data_dir, "merged/gene_count_matrix.csv"))
count_mat = rename(count_mat, c("V1"="gene_id"))  # rename for convenience

# Gene annotation table matching the count matrix
gene_annot = data.frame(gene_id=count_mat$gene_id)

# Map gene IDs to the merged GTF file specifiying trancsripts detected across samples
idx = match(gene_annot$gene_id, read_annot$gene_id)

message("Gene IDs without match: ", sum(is.na(idx)))

# Transfer to annotation table
gene_annot$gene_name = read_annot$gene_name[idx]
gene_annot$ref_gene_id = read_annot$ref_gene_id[idx]

# gene_name = read_annot$gene_name[idx]
message("Transcripts without gene names: ", sum(is.na(gene_annot$gene_name)))

write.csv(gene_annot,
	file.path(data_dir, "merged/gene_annot.csv"),
	row.names=FALSE,
	quote=FALSE)